cdap-app-fabric/src/main/java/co/cask/cdap/route/store/ZKRouteStore.java

```diff
@@ -154,13 +155,48 @@ public void onSuccess(NodeData result) {
 
       @Override
       public void onFailure(Throwable t) {
-        routeConfigMap.remove(serviceId);
         settableFuture.setException(t);
+        // If no node is present, then simply place a watch on the node to monitor further data updates
+        // If its an exception other than NoNodeException, then remove the future, so that zkClient#getData
+        // will be retried during the next fetch config request
+        if (t instanceof KeeperException.NoNodeException) {
+          existsAndWatch(serviceId);
+        } else {
+          routeConfigMap.remove(serviceId);
```

Comment:
That means any disconnection to ZK from a given router instance will make the routing super slow even we have the list of discoverable to route to? Shouldn't we fallback to the default handling (configurable strategy)?